lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp

```diff
@@ -193,8 +223,317 @@ struct LowerClassesPass : public LowerClassesBase<LowerClassesPass> {
+              const DenseMap<DistinctAttr, FModuleOp> &owningModules)
+      : module(module), moduleNamespace(moduleNamespace),
+        instanceGraph(instanceGraph), symbolTable(symbolTable),
+        owningModules(owningModules) {}
+
+  struct PathInfoTableEntry {
+    Operation *op;
+    DistinctAttr id;
+    StringAttr altBasePathModule;
+    // This is null if the path has no owning module.
+    ArrayAttr pathAttr;
+  };
+
+  LogicalResult runOnModule();
+  LogicalResult processPathTrackers(AnnoTarget target);
+  LogicalResult updatePathInfoTable(PathInfoTable &pathInfoTable,
+                                    HierPathCache &cache);
+
+  // Determine it is necessary to use an alternative base path for `moduleName`
+  // and `owningModule`.
+  FailureOr<bool> getOrComputeNeedsAltBasePath(Location loc,
+                                               StringAttr moduleName,
+                                               FModuleOp owningModule);
+  FModuleLike module;
+
+  // Local data structures.
+  hw::InnerSymbolNamespace &moduleNamespace;
+  DenseMap<std::pair<StringAttr, FModuleOp>, bool> needsAltBasePathCache;
+  SmallVector<Attribute> portAnnotations;
+
+  // Thread-unsafe global data structure. Don't mutate.
+  InstanceGraph &instanceGraph;
+  const SymbolTable &symbolTable;
+  const DenseMap<DistinctAttr, FModuleOp> &owningModules;
+
+  // Result.
+  SmallVector<PathInfoTableEntry> entries;
+  SetVector<StringAttr> altBasePathRoots;
+};
+
 } // namespace
 
+LogicalResult
+PathTracker::run(CircuitOp circuit, InstanceGraph &instanceGraph,
+                 hw::InnerSymbolNamespaceCollection &namespaces,
+                 HierPathCache &cache, PathInfoTable &pathInfoTable,
+                 const SymbolTable &symbolTable,
+                 const DenseMap<DistinctAttr, FModuleOp> &owningModules) {
+  SmallVector<PathTracker> trackers;
+  // Prepare workers.
+  for (auto module : circuit.getOps<FModuleLike>())
+    trackers.push_back(PathTracker(module, namespaces[module], instanceGraph,
+                                   symbolTable, owningModules));
+
+  if (failed(failableParallelForEach(
+          circuit.getContext(), trackers,
+          [](PathTracker &tracker) { return tracker.runOnModule(); })))
+    return failure();
+
+  // Update the pathInfoTable sequentially.
+  for (auto &tracker : trackers)
+    if (failed(tracker.updatePathInfoTable(pathInfoTable, cache)))
+      return failure();
+
+  return success();
+}
+
+LogicalResult PathTracker::runOnModule() {
+  // Process the module annotations.
+  if (failed(processPathTrackers(OpAnnoTarget(module))))
+    return failure();
+  // Process module port annotations.
+  for (unsigned i = 0, e = module.getNumPorts(); i < e; ++i)
+    if (failed(processPathTrackers(PortAnnoTarget(module, i))))
+      return failure();
+  module->setAttr("portAnnotations",
+                  ArrayAttr::get(module.getContext(), portAnnotations));
```

Comment:
is there a `setPortAnnotationsAttr` that we can use?  Otherwise maybe `PortAnnoTarget::setAnnotations(..)`?

Reply:
Seems like missing method for FModuleLike. Added setPortAnnotationsAttr.